src/watchdog.js

```diff
@@ -0,0 +1,393 @@
+/**
+ * @license Copyright (c) 2003-2019, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+/**
+ * @module watchdog/watchdog
+ */
+
+/* globals console, window */
+
+import mix from '@ckeditor/ckeditor5-utils/src/mix';
+import EmitterMixin from '@ckeditor/ckeditor5-utils/src/emittermixin';
+import { throttle } from 'lodash-es';
+import CKEditorError from '@ckeditor/ckeditor5-utils/src/ckeditorerror';
+import areConnectedThroughProperties from '@ckeditor/ckeditor5-utils/src/areconnectedthroughproperties';
+
+/**
+ * A Watchdog for CKEditor 5 editors.
+ *
+ * It keeps the {@link module:core/editor/editor~Editor editor} instance running. If some error occurs in the editor it tries to
+ * restart it to the previous state.
+ *
+ * It does not handle errors during editor initialization and editor destruction.
+ *
+ * Basic usage:
+ *
+ * 		const watchdog = Watchdog.for( ClassicEditor );
+ *
+ * 		watchdog.create( elementOrData, editorConfig ).then( editor => {} );
```

Comment:
Doesn't `watchdog.create()` sound like it creates a watchdog?

You have `restart()` there, so one option could be to call this method `start()`.

The other option might be to present it like this:

```js
const watchdoggedEditor = Watchdog.for( ClassicEditor );

watchdoggedEditor.create().then();
```

But I think that this wouldn't be correct because this `watchdoggedEditor` isn't an editor. So I'd rather use `start()`.

Reply:
To be honest, to do not feel this difference between `create` and `start`. In fact, for me it is the opposite:
 - `watchdog.create` sounds like it creates something else, you do not create yourself. Like artists who create a painting, the painting does not create itself. The same way you ask watchdog to create the editor, 
 - `watchdog.start` sounds like it is `watchdog` which starts. When you press a "start" button on the coffee machine, you start the coffee machine, you do not ask it to start something else.

So, for me, there could be `editor.start` and `watchdog.create`, but I do not think it is a good moment to change it.

I also like the fact that `watchdog` and `editor` has a quite similar API, it could make using it even transparent in some simple cases.

Reply:
For me, the `start` would be fine it would just start watching the editor. But it creates it and starts watching instead. Similarly, there's `watchdog.destroy()` instead of `watchdog.stop()`.

Reply:
I mean: I am fine with both `create` and `start`. The only real argument for `create` is to make the API more transparent. Note, that we may add watchdog to integrations one day, and it might be tricky how methods there should be called if watchdog call methods differently then the editor.